### ISSUE TITLE
TCP simultaneous open don't work

### ISSUE THREAD

**Issue by rhornig** [inet-framework/inet#92]
Two active open for each other not worked correctly:
recorded states: CLOSED -> SYN_SENT -> SYN_RECEIVED -> CLOSED
expected states: CLOSED -> SYN_SENT -> SYN_RECEIVED -> ESTABLISHED

The received SYN+ACK packet refused in SYN_RECEIVED status, state changed to Closed instead ESTABLISHED

See the test in INET/tests/module/tcp_syn_2.test.off:
problem occured in step 6 and step 7 in RFC 793, figure 8.
Additional Information  RFC 793, page 32: (fixed by RFC 1122)
Simultaneous initiation is only slightly more complex, as is shown in figure 8.
Each TCP cycles from CLOSED to SYN-SENT to SYN-RECEIVED to ESTABLISHED.

```
  TCP A TCP B
```
1. CLOSED CLOSED
2. SYN-SENT --> <SEQ=100><CTL=SYN> ...
3. SYN-RECEIVED <-- <SEQ=300><CTL=SYN> <-- SYN-SENT
4. ... <SEQ=100><CTL=SYN> --> SYN-RECEIVED
5. SYN-RECEIVED --> <SEQ=100><ACK=301><CTL=SYN,ACK> ...
6. ESTABLISHED <-- <SEQ=300><ACK=101><CTL=SYN,ACK> <-- SYN-RECEIVED
7. ... <SEQ=100><ACK=301><CTL=ACK> --> ESTABLISHED
   
   ```
           Simultaneous Connection Synchronization
   
                          Figure 8.
   ```

from: http://dev.omnetpp.org/bugs/view.php?id=616

**Comment by avarga** [comment]
In the case of simultaneous open, the state transitions of either TCP entity should be:  `SYN_SENT` --(upon receiving SYN)--> `SYN_RECEIVED` --(upon receiving SYN+ACK)--> `ESTABLISHED`.

It is the 2nd transition, `SYN_RECEIVED` --> `ESTABLISHED` which goes wrong. The SYN is treated as an unexpected one, and the connection closes.

This is due to an omission in RFC 793 (as the INET TCP code closely follows the RFC).  The description of the procedure to be followed when a segment arrives starts on Page 65. It begins with special cases: `If the state is CLOSED`,  `If the state is LISTEN`, `If the state is SYN-SENT`, then goes on to `Otherwise` for all other states. 

For our case, the `Otherwise` branch is relevant. After some (for us) irrelevant bits, it says:

    fourth, check the SYN bit,

      SYN-RECEIVED
      ESTABLISHED STATE
      FIN-WAIT STATE-1
      FIN-WAIT STATE-2
      CLOSE-WAIT STATE
      CLOSING STATE
      LAST-ACK STATE
      TIME-WAIT STATE

        If the SYN is in the window it is an error, send a reset, any
        outstanding RECEIVEs and SEND should receive "reset" responses,
        all segment queues should be flushed, the user should also
        receive an unsolicited general "connection reset" signal, enter
        the CLOSED state, delete the TCB, and return.

        If the SYN is not in the window this step would not be reached
        and an ack would have been sent in the first step (sequence
        number check).

Which says nothing about going to `ESTABLISHED` if the SYN (SYN+ACK) is received in the `SYN_RECEIVED` state. The INET TCP code, being compliant, faithfully went to the `CLOSED` state.

This is clearly not what's supposed to be done. For example, lwIP's implementation contains this:

        tcp_process(struct tcp_pcb *pcb)
        {
            ...
            switch (pcb->state) {
            case SYN_RCVD:
                if (flags & TCP_ACK) {
                /* expected ACK number? */
                if (TCP_SEQ_BETWEEN(ackno, pcb->lastack+1, pcb->snd_nxt)) {
                    u16_t old_cwnd;
                    pcb->state = ESTABLISHED;
            ...
        }

So it treats being in the SYN_RCVD state explicitly.

As the fix, our procedure in the INET code for the `fourth, check the SYN bit,` section of the RFC was modified as it it was protected by a `state != SYN_RCVD` condition.

**Comment by avarga** [comment]
Fixed in INET 3.8, INET 4.4
